In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## Kobe Bryant  shot selection

Cсылка на соревнование: https://www.kaggle.com/c/kobe-bryant-shot-selection

Goal: Fun and education

Using 20 years of data on Kobe's swishes and misses, can you predict which shots will find the bottom of the net? This competition is well suited for practicing classification basics, feature engineering, and time series analysis. Practice got Kobe an eight-figure contract and 5 championship rings. What will it get you?

This data contains the location and circumstances of every field goal attempted by Kobe Bryant took during his 20-year career. Your task is to predict whether the basket went in (shot_made_flag).

We have removed 5000 of the shot_made_flags (represented as missing values in the csv file). These are the test set shots for which you must submit a prediction. You are provided a sample submission file with the correct shot_ids needed for a valid prediction.

In [2]:
data = pd.read_csv('Kobe.csv')

In [3]:
data.head()

action_type combined_shot_type  game_event_id   game_id      lat  \
0          Jump Shot          Jump Shot             10  20000012  33.9723   
1          Jump Shot          Jump Shot             12  20000012  34.0443   
2          Jump Shot          Jump Shot             35  20000012  33.9093   
3          Jump Shot          Jump Shot             43  20000012  33.8693   
4  Driving Dunk Shot               Dunk            155  20000012  34.0443   

   loc_x  loc_y       lon  minutes_remaining  period   ...          shot_type  \
0    167     72 -118.1028                 10       1   ...     2PT Field Goal   
1   -157      0 -118.4268                 10       1   ...     2PT Field Goal   
2   -101    135 -118.3708                  7       1   ...     2PT Field Goal   
3    138    175 -118.1318                  6       1   ...     2PT Field Goal   
4      0      0 -118.2698                  6       2   ...     2PT Field Goal   

          shot_zone_area  shot_zone_basic  shot_zone_range     team_id  \
0          Right Side(R)        Mid-Range        16-24 ft.  1610612747   
1           Left Side(L)        Mid-Range         8-16 ft.  1610612747   
2   Left Side Center(LC)        Mid-Range        16-24 ft.  1610612747   
3  Right Side Center(RC)        Mid-Range        16-24 ft.  1610612747   
4              Center(C)  Restricted Area  Less Than 8 ft.  1610612747   

            team_name   game_date    matchup opponent  shot_id  
0  Los Angeles Lakers  2000-10-31  LAL @ POR      POR        1  
1  Los Angeles Lakers  2000-10-31  LAL @ POR      POR        2  
2  Los Angeles Lakers  2000-10-31  LAL @ POR      POR        3  
3  Los Angeles Lakers  2000-10-31  LAL @ POR      POR        4  
4  Los Angeles Lakers  2000-10-31  LAL @ POR      POR        5  

[5 rows x 25 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30697 entries, 0 to 30696
Data columns (total 25 columns):
action_type           30697 non-null object
combined_shot_type    30697 non-null object
game_event_id         30697 non-null int64
game_id               30697 non-null int64
lat                   30697 non-null float64
loc_x                 30697 non-null int64
loc_y                 30697 non-null int64
lon                   30697 non-null float64
minutes_remaining     30697 non-null int64
period                30697 non-null int64
playoffs              30697 non-null int64
season                30697 non-null object
seconds_remaining     30697 non-null int64
shot_distance         30697 non-null int64
shot_made_flag        25697 non-null float64
shot_type             30697 non-null object
shot_zone_area        30697 non-null object
shot_zone_basic       30697 non-null object
shot_zone_range       30697 non-null object
team_id               30697 non-null int64
team_name         

пропущенные значения есть только в таргете, которые мы должны предсказать

In [5]:
data['combined_shot_type'].value_counts()

Jump Shot    23485
Layup         5448
Dunk          1286
Tip Shot       184
Hook Shot      153
Bank Shot      141
Name: combined_shot_type, dtype: int64

In [6]:
data_for_an=data.dropna()

In [7]:
data_for_an.shape

(25697, 25)

 Описание данных
 Группа 1.Координаты
 -широта/долгота
 -loc_x/loc_y
 -shot_distance
 -shot_type
 Координаты географические и декартовы по сути содержат одну и ту же информацию-место точки удара на карте.
 Можно и нужно (для линейной регрессии) выбрать что-то одно, что по каким-то причинам будет работать лучше.
 shot distance и shot type то же одно и тоже- двух-или трехочковый бросок зависит от дальности.
 Информация об этом тоже содержится в первых двух переменных. Только еще в первых двух переменных содержится информация об угле.
 Группа 2.Переменные для группировки координат
 -shot_zone_area
 -shot_zone_basic
 -shot_zone_range
 В чем-то содержат пересекающуюся информацию, в чем-то разную. Характеризуют область на поле. Можно сгруппировать по ним координаты.
 Группа 3 Сезон
 - из даты игры вытянуть месяц - чтобы понять, если зависимость от части сезона (начало или конец)
 - season- года- посмотреть количество бросков на количество игр по сезонам.
 Группа 4 Время в игре
 Период+минуты (начало периода/середина/конец)
 Группа 3 (месяц) и 4 обработать как качественные переменные
 Группа 5. Тип броска
 action_type и combined_shot_type Лучше использовать второе (качественная переменная)

Проверим на выбросы и разделим на тест и трейн

In [8]:
data_for_an.describe(percentiles=(0.01, 0.99))

game_event_id       game_id           lat         loc_x         loc_y  \
count   25697.000000  2.569700e+04  25697.000000  25697.000000  25697.000000   
mean      249.348679  2.474109e+07     33.953043      7.148422     91.257345   
std       149.778520  7.738108e+06      0.088152    110.073147     88.152106   
min         2.000000  2.000001e+07     33.253300   -250.000000    -44.000000   
1%          5.000000  2.000021e+07     33.773300   -225.000000    -18.000000   
50%       253.000000  2.090034e+07     33.970300      0.000000     74.000000   
99%       545.000000  4.990003e+07     34.062300    228.000000    271.000000   
max       653.000000  4.990009e+07     34.088300    248.000000    791.000000   

                lon  minutes_remaining        period      playoffs  \
count  25697.000000       25697.000000  25697.000000  25697.000000   
mean    -118.262652           4.886796      2.520800      0.146243   
std        0.110073           3.452475      1.151626      0.353356   
min     -118.519800           0.000000      1.000000      0.000000   
1%      -118.494800           0.000000      1.000000      0.000000   
50%     -118.269800           5.000000      3.000000      0.000000   
99%     -118.041800          11.000000      5.000000      1.000000   
max     -118.021800          11.000000      7.000000      1.000000   

       seconds_remaining  shot_distance  shot_made_flag       team_id  \
count       25697.000000   25697.000000    25697.000000  2.569700e+04   
mean           28.311554      13.457096        0.446161  1.610613e+09   
std            17.523392       9.388725        0.497103  0.000000e+00   
min             0.000000       0.000000        0.000000  1.610613e+09   
1%              0.000000       0.000000        0.000000  1.610613e+09   
50%            28.000000      15.000000        0.000000  1.610613e+09   
99%            59.000000      28.000000        1.000000  1.610613e+09   
max            59.000000      79.000000        1.000000  1.610613e+09   

            shot_id  
count  25697.000000  
mean   15328.166946  
std     8860.462397  
min        2.000000  
1%       310.960000  
50%    15336.000000  
99%    30378.080000  
max    30697.000000

Из переменных для анализа loc_y имеет выброс и shot_distance

In [9]:
data_for_an=data_for_an[data_for_an['loc_y']<272]

In [10]:
data_for_an.describe(percentiles=(0.01, 0.99))

game_event_id       game_id           lat         loc_x         loc_y  \
count   25442.000000  2.544200e+04  25442.000000  25442.000000  25442.000000   
mean      248.897257  2.473374e+07     33.955879      7.236459     88.421468   
std       149.776246  7.731662e+06      0.082943    110.138580     82.943264   
min         2.000000  2.000001e+07     33.773300   -250.000000    -44.000000   
1%          5.000000  2.000021e+07     33.785300   -225.000000    -18.000000   
50%       253.000000  2.090033e+07     33.973300      0.000000     71.000000   
99%       544.000000  4.990003e+07     34.062300    228.000000    259.000000   
max       653.000000  4.990009e+07     34.088300    248.000000    271.000000   

                lon  minutes_remaining        period      playoffs  \
count  25442.000000       25442.000000  25442.000000  25442.000000   
mean    -118.262564           4.921193      2.520439      0.146136   
std        0.110139           3.441681      1.151813      0.353250   
min     -118.519800           0.000000      1.000000      0.000000   
1%      -118.494800           0.000000      1.000000      0.000000   
50%     -118.269800           5.000000      3.000000      0.000000   
99%     -118.041800          11.000000      5.000000      1.000000   
max     -118.021800          11.000000      7.000000      1.000000   

       seconds_remaining  shot_distance  shot_made_flag       team_id  \
count       25442.000000   25442.000000    25442.000000  2.544200e+04   
mean           28.484985      13.208474        0.449139  1.610613e+09   
std            17.447462       9.009959        0.497416  0.000000e+00   
min             0.000000       0.000000        0.000000  1.610613e+09   
1%              0.000000       0.000000        0.000000  1.610613e+09   
50%            28.000000      15.000000        0.000000  1.610613e+09   
99%            59.000000      27.000000        1.000000  1.610613e+09   
max            59.000000      31.000000        1.000000  1.610613e+09   

            shot_id  
count  25442.000000  
mean   15321.976889  
std     8855.274562  
min        2.000000  
1%       311.410000  
50%    15318.500000  
99%    30376.590000  
max    30697.000000

In [11]:
from sklearn.model_selection import train_test_split

data_train, data_test = train_test_split(data_for_an, test_size=0.2)
len(data_train)

20353

In [12]:
groups_coordinates=data_train.groupby(['shot_zone_area','shot_zone_basic','shot_zone_range']).mean()[['loc_x','loc_y']]

In [13]:
groups_coordinates

loc_x  \
shot_zone_area        shot_zone_basic       shot_zone_range               
Center(C)             Above the Break 3     24+ ft.           -0.015625   
                      In The Paint (Non-RA) 8-16 ft.           3.659138   
                                            Less Than 8 ft.    2.254851   
                      Mid-Range             16-24 ft.          0.717822   
                                            8-16 ft.          -1.443946   
                      Restricted Area       Less Than 8 ft.    0.371435   
Left Side Center(LC)  Above the Break 3     24+ ft.         -161.165636   
                      Mid-Range             16-24 ft.       -114.368760   
Left Side(L)          In The Paint (Non-RA) 8-16 ft.         -66.164659   
                      Left Corner 3         24+ ft.         -232.494565   
                      Mid-Range             16-24 ft.       -173.895181   
                                            8-16 ft.        -118.210227   
Right Side Center(RC) Above the Break 3     24+ ft.          154.468321   
                      Mid-Range             16-24 ft.        113.725200   
Right Side(R)         In The Paint (Non-RA) 8-16 ft.          68.580060   
                      Mid-Range             16-24 ft.        172.836364   
                                            8-16 ft.         119.704689   
                      Right Corner 3        24+ ft.          231.269784   

                                                                  loc_y  
shot_zone_area        shot_zone_basic       shot_zone_range              
Center(C)             Above the Break 3     24+ ft.          252.320312  
                      In The Paint (Non-RA) 8-16 ft.         107.603696  
                                            Less Than 8 ft.   43.809185  
                      Mid-Range             16-24 ft.        191.002475  
                                            8-16 ft.         148.372197  
                      Restricted Area       Less Than 8 ft.    3.117050  
Left Side Center(LC)  Above the Break 3     24+ ft.          196.080412  
                      Mid-Range             16-24 ft.        151.607890  
Left Side(L)          In The Paint (Non-RA) 8-16 ft.          75.550201  
                      Left Corner 3         24+ ft.           19.994565  
                      Mid-Range             16-24 ft.         51.150602  
                                            8-16 ft.          41.158279  
Right Side Center(RC) Above the Break 3     24+ ft.          199.064664  
                      Mid-Range             16-24 ft.        153.232902  
Right Side(R)         In The Paint (Non-RA) 8-16 ft.          76.060423  
                      Mid-Range             16-24 ft.         42.897608  
                                            8-16 ft.          35.567530  
                      Right Corner 3        24+ ft.           17.374101

In [14]:
data_train = pd.merge(data_train, groups_coordinates,  how='left', left_on=['shot_zone_area', 'shot_zone_basic', 'shot_zone_range'], right_on = ['shot_zone_area', 'shot_zone_basic', 'shot_zone_range'])

In [15]:
data_test = pd.merge(data_test, groups_coordinates,  how='left', left_on=['shot_zone_area', 'shot_zone_basic', 'shot_zone_range'], right_on = ['shot_zone_area', 'shot_zone_basic', 'shot_zone_range'])

In [16]:
from datetime import datetime
data_1=[]
for i in data_train['game_date'].values:
    a=datetime.strptime(i, '%Y-%m-%d')
    a=a.month
    data_1.append(a)
data_train['month']=data_1
data_2=[]
for i in data_test['game_date'].values:
    a=datetime.strptime(i, '%Y-%m-%d')
    a=a.month
    data_2.append(a)
data_test['month']=data_2

In [17]:
data_season=data_train.groupby(['season','game_id']).count()

In [18]:
data_season=data_season.groupby('season').count()[['action_type','lat']]

In [19]:
data_season_1=data_train.groupby(['season']).count()[['action_type','lat']]

In [20]:
data_season=data_season.merge(data_season_1, how='left', left_on='season', right_on='season')
data_season['shot_per_game']=data_season['lat_y']/data_season['lat_x']
data_season['shots']=data_season['lat_y']
data_season=data_season[['shot_per_game','shots']]

In [21]:
data_train = pd.merge(data_train, data_season,  how='left', left_on=['season'], right_on = ['season'])
data_test = pd.merge(data_test, data_season,  how='left', left_on=['season'], right_on = ['season'])

In [22]:
def minutes_category(db):
    minutes_cat=[]
    for i in db['minutes_remaining'].values:
        if i<=4:
            minutes_cat.append('finish')
        else:
            if i<=8:
                minutes_cat.append('middle')  
            else:
                minutes_cat.append('start')
    print (len(minutes_cat))
    db['minutes_cat']=minutes_cat

In [23]:
minutes_category(data_train)
minutes_category(data_test)

20353
5089


In [24]:
period_str=[]
for i in data_train['period']:
    i=str(i)
    period_str.append(i)

In [25]:
data_train['period_str']=period_str
data_train['period_min']=data_train['period_str']+data_train['minutes_cat']

In [26]:
period_str=[]
for i in data_test['period']:
    i=str(i)
    period_str.append(i)
data_test['period_str']=period_str
data_test['period_min']=data_test['period_str']+data_test['minutes_cat']

In [27]:
data_train['locxLocy']=data_train['loc_x_y']*data_train['loc_y_y']
data_test['locxLocy']=data_test['loc_x_y']*data_test['loc_y_y']

In [28]:
data_train_for_an=data_train[['combined_shot_type','loc_x_y','loc_y_y','locxLocy','month','shot_per_game', 'period_min','shot_made_flag']]

In [29]:
data_test_for_an=data_test[['combined_shot_type','loc_x_y','loc_y_y','locxLocy','month','shot_per_game', 'period_min','shot_made_flag']]

In [30]:
def get_woe_v1(db, db1, col, target_col):
    all_good = len(db[db[target_col] == 1][col])
    all_bad = len(db[db[target_col] == 0][col])
    odds_series = (
        db[db[target_col] == 1][col].value_counts()
        /
        db[db[target_col] == 0][col].value_counts()
    )
    odds_series = odds_series / all_good * all_bad
    category_woe_dict = np.log(odds_series).to_dict()
    db[col + '_woe'] = db[col].apply(category_woe_dict.get)
    db1[col + '_woe'] = db1[col].apply(category_woe_dict.get)
    return db, db1

In [31]:
res=['combined_shot_type', 'month', 'period_min']
for i in res:
    get_woe_v1(data_train_for_an, data_test_for_an, i, 'shot_made_flag')


C:\Users\volko\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # This is added back by InteractiveShellApp.init_path()
C:\Users\volko\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if sys.path[0] == '':


In [32]:
data_train_for_an=data_train_for_an.drop(res, axis=1)
data_test_for_an=data_test_for_an.drop(res, axis=1)

In [33]:
corr_matrix = data_train_for_an.corr()
corr_matrix

loc_x_y   loc_y_y  locxLocy  shot_per_game  \
loc_x_y                 1.000000 -0.017881  0.824711       0.019465   
loc_y_y                -0.017881  1.000000  0.026807       0.033672   
locxLocy                0.824711  0.026807  1.000000       0.014318   
shot_per_game           0.019465  0.033672  0.014318       1.000000   
shot_made_flag          0.002028 -0.139402  0.010310       0.011653   
combined_shot_type_woe -0.025726 -0.422514 -0.014231      -0.041827   
month_woe              -0.008983 -0.003341 -0.006911       0.007566   
period_min_woe          0.002697 -0.055690  0.006116       0.010417   

                        shot_made_flag  combined_shot_type_woe  month_woe  \
loc_x_y                       0.002028               -0.025726  -0.008983   
loc_y_y                      -0.139402               -0.422514  -0.003341   
locxLocy                      0.010310               -0.014231  -0.006911   
shot_per_game                 0.011653               -0.041827   0.007566   
shot_made_flag                1.000000                0.241270   0.016648   
combined_shot_type_woe        0.241270                1.000000   0.005813   
month_woe                     0.016648                0.005813   1.000000   
period_min_woe                0.052403                0.011349   0.003521   

                        period_min_woe  
loc_x_y                       0.002697  
loc_y_y                      -0.055690  
locxLocy                      0.006116  
shot_per_game                 0.010417  
shot_made_flag                0.052403  
combined_shot_type_woe        0.011349  
month_woe                     0.003521  
period_min_woe                1.000000

In [34]:
from sklearn.linear_model import Lasso
from sklearn.metrics import roc_auc_score

def lasso_regression(table, predictors, alpha):
    lassoreg = Lasso(alpha=alpha, normalize=True, max_iter=1e5)
    lassoreg.fit(predictors, table['shot_made_flag'])
    y_pred = lassoreg.predict(predictors)
    roc_auc = np.round(roc_auc_score(table['shot_made_flag'], y_pred), 2)
    rss = sum((y_pred-table['shot_made_flag']) ** 2)
    ret = [rss, roc_auc]
    ret.extend([lassoreg.intercept_])
    ret.extend(lassoreg.coef_)
    return ret

In [35]:
predictors=data_train_for_an.drop('shot_made_flag', axis=1)
predictors1=data_test_for_an.drop('shot_made_flag', axis=1)

In [36]:
alpha_lasso = [1e-15, 1e-10, 1e-8, 1e-5,1e-4, 1e-3,1e-2, 1, 5, 10]

# create dataframe to store coefficients
col = ['rss','roc_auc', 'intercept'] + predictors.columns.tolist()
ind = ['alpha_%.2g' % alpha_lasso[i] for i in range(0, 10)]
coef_matrix_lasso = pd.DataFrame(index=ind, columns=col)

#Iterate over the 10 alpha values:
for i in range(10):
    coef_matrix_lasso.iloc[i, ] = lasso_regression(data_train_for_an, predictors, alpha_lasso[i])

In [37]:
coef_matrix_lasso

rss roc_auc intercept      loc_x_y      loc_y_y  \
alpha_1e-15   4718.98    0.61  0.406597 -7.48553e-05 -0.000282504   
alpha_1e-10   4718.98    0.61  0.406597 -7.48545e-05 -0.000282504   
alpha_1e-08   4718.98    0.61  0.406604 -7.47771e-05 -0.000282485   
alpha_1e-05   4719.61    0.61  0.413577           -0 -0.000264154   
alpha_0.0001  4733.53    0.61   0.45839            0 -0.000154144   
alpha_0.001   5037.08     0.5   0.44986            0           -0   
alpha_0.01    5037.08     0.5   0.44986            0           -0   
alpha_1       5037.08     0.5   0.44986            0           -0   
alpha_5       5037.08     0.5   0.44986            0           -0   
alpha_10      5037.08     0.5   0.44986            0           -0   

                 locxLocy shot_per_game combined_shot_type_woe month_woe  \
alpha_1e-15    9.7751e-07    0.00456015               0.173019    0.2217   
alpha_1e-10   9.77504e-07    0.00456014               0.173019    0.2217   
alpha_1e-08   9.76912e-07     0.0045595               0.173018  0.221659   
alpha_1e-05   4.00884e-07    0.00391898               0.171816  0.180328   
alpha_0.0001            0             0               0.156977         0   
alpha_0.001             0             0                      0         0   
alpha_0.01              0             0                      0         0   
alpha_1                 0             0                      0         0   
alpha_5                 0             0                      0         0   
alpha_10                0             0                      0         0   

             period_min_woe  
alpha_1e-15        0.219364  
alpha_1e-10        0.219364  
alpha_1e-08        0.219352  
alpha_1e-05        0.207265  
alpha_0.0001      0.0937367  
alpha_0.001               0  
alpha_0.01                0  
alpha_1                   0  
alpha_5                   0  
alpha_10                  0

Обнуляются месяц, среднее количество ударов за игру, координата X и совокупная координата

In [38]:
# create dataframe to store coefficients
col = ['rss','roc_auc', 'intercept'] + predictors.columns.tolist()
ind = ['alpha_%.2g' % alpha_lasso[i] for i in range(0, 10)]
coef_matrix_lasso_test = pd.DataFrame(index=ind, columns=col)

#Iterate over the 10 alpha values:
for i in range(10):
    coef_matrix_lasso_test.iloc[i, ] = lasso_regression(data_test_for_an, predictors1, alpha_lasso[i])
coef_matrix_lasso_test

rss roc_auc intercept      loc_x_y      loc_y_y  \
alpha_1e-15   1177.74    0.61  0.408694  1.13971e-05 -0.000335256   
alpha_1e-10   1177.74    0.61  0.408694  1.13967e-05 -0.000335256   
alpha_1e-08   1177.74    0.61  0.408698  1.13587e-05 -0.000335254   
alpha_1e-05   1177.75    0.61  0.412405  3.40482e-06 -0.000329659   
alpha_0.0001  1178.78    0.61  0.445701            0 -0.000277446   
alpha_0.001   1207.89     0.6  0.443468           -0           -0   
alpha_0.01    1257.55     0.5  0.446257           -0           -0   
alpha_1       1257.55     0.5  0.446257           -0           -0   
alpha_5       1257.55     0.5  0.446257           -0           -0   
alpha_10      1257.55     0.5  0.446257           -0           -0   

                 locxLocy shot_per_game combined_shot_type_woe month_woe  \
alpha_1e-15  -1.09811e-08    0.00444394               0.174254  0.169172   
alpha_1e-10  -1.09782e-08    0.00444394               0.174254  0.169172   
alpha_1e-08   -1.0688e-08    0.00444364               0.174253   0.16915   
alpha_1e-05             0    0.00414243               0.173344  0.147605   
alpha_0.0001            0    0.00139355               0.165557         0   
alpha_0.001            -0             0              0.0791623         0   
alpha_0.01             -0             0                      0         0   
alpha_1                -0             0                      0         0   
alpha_5                -0             0                      0         0   
alpha_10               -0             0                      0         0   

             period_min_woe  
alpha_1e-15       0.0875461  
alpha_1e-10        0.087546  
alpha_1e-08         0.08754  
alpha_1e-05       0.0815808  
alpha_0.0001      0.0277549  
alpha_0.001               0  
alpha_0.01                0  
alpha_1                   0  
alpha_5                   0  
alpha_10                  0

In [39]:
from sklearn.linear_model import Ridge

def ridge_regression(data, predictors, alpha):
    ridgereg = Ridge(alpha=alpha, normalize=True)
    ridgereg.fit(predictors, data['shot_made_flag'])
    y_pred = ridgereg.predict(predictors)
    rss = sum((y_pred-data['shot_made_flag']) ** 2)
    roc_auc = np.round(roc_auc_score(data['shot_made_flag'], y_pred), 2)
    ret = [rss, roc_auc]
    ret.extend([ridgereg.intercept_])
    ret.extend(ridgereg.coef_)
    return ret

In [40]:
alpha_ridge = [1e-15, 1e-10, 1e-8, 1e-4, 1e-3, 1e-2, 1, 5, 10, 20]
col = ['rss','roc_auc', 'intercept'] + predictors.columns.tolist()
ind = ['alpha_%.2g' % alpha_ridge[i] for i in range(0, 10)]
coef_matrix_ridge = pd.DataFrame(index=ind, columns=col)

for i in range(10):
    coef_matrix_ridge.iloc[i, ] = ridge_regression(data_train_for_an, predictors, alpha_ridge[i])

In [41]:
coef_matrix_ridge

rss roc_auc intercept      loc_x_y      loc_y_y  \
alpha_1e-15   4718.98    0.61  0.406597 -7.48553e-05 -0.000282504   
alpha_1e-10   4718.98    0.61  0.406597 -7.48553e-05 -0.000282504   
alpha_1e-08   4718.98    0.61  0.406597 -7.48553e-05 -0.000282504   
alpha_0.0001  4718.98    0.61   0.40661 -7.48071e-05  -0.00028254   
alpha_0.001   4718.98    0.61   0.40672 -7.43764e-05 -0.000282862   
alpha_0.01    4719.01    0.61  0.407807 -7.02971e-05 -0.000285995   
alpha_1       4785.55    0.61  0.447717 -3.98106e-06  -0.00029174   
alpha_5        4922.3    0.61  0.453756  8.36976e-07 -0.000130208   
alpha_10      4969.57    0.61  0.452608  6.51723e-07 -7.53424e-05   
alpha_20      5000.13    0.61  0.451492  3.94955e-07 -4.07903e-05   

                 locxLocy shot_per_game combined_shot_type_woe  month_woe  \
alpha_1e-15    9.7751e-07    0.00456015               0.173019     0.2217   
alpha_1e-10    9.7751e-07    0.00456015               0.173019     0.2217   
alpha_1e-08    9.7751e-07    0.00456015               0.173019     0.2217   
alpha_0.0001  9.77104e-07    0.00455952                  0.173   0.221681   
alpha_0.001   9.73466e-07    0.00455388               0.172828   0.221505   
alpha_0.01     9.3887e-07    0.00449821               0.171131   0.219762   
alpha_1       2.42451e-07    0.00182615              0.0862139   0.116929   
alpha_5       6.65104e-08   0.000482344              0.0301115  0.0404223   
alpha_10       3.4939e-08   0.000244992              0.0166806  0.0222441   
alpha_20       1.7916e-08   0.000122651             0.00882212   0.011712   

             period_min_woe  
alpha_1e-15        0.219364  
alpha_1e-10        0.219364  
alpha_1e-08        0.219364  
alpha_0.0001       0.219342  
alpha_0.001        0.219147  
alpha_0.01         0.217214  
alpha_1            0.113028  
alpha_5           0.0394999  
alpha_10          0.0218414  
alpha_20           0.011536

минимальное количество переменных без снижения качества достигается при alpha_1. Сильнее всего близки к 0 координата x, скомбинированная x-y

In [42]:
alpha_ridge = [1e-15, 1e-10, 1e-8, 1e-4, 1e-3, 1e-2, 1, 5, 10, 20]
col = ['rss','roc_auc', 'intercept'] + predictors.columns.tolist()
ind = ['alpha_%.2g' % alpha_ridge[i] for i in range(0, 10)]
coef_matrix_ridge1 = pd.DataFrame(index=ind, columns=col)

for i in range(10):
    coef_matrix_ridge1.iloc[i, ] = ridge_regression(data_test_for_an, predictors1, alpha_ridge[i])

In [43]:
coef_matrix_ridge1

rss roc_auc intercept      loc_x_y      loc_y_y  \
alpha_1e-15   1177.74    0.61  0.408694  1.13971e-05 -0.000335256   
alpha_1e-10   1177.74    0.61  0.408694  1.13971e-05 -0.000335256   
alpha_1e-08   1177.74    0.61  0.408694  1.13971e-05 -0.000335256   
alpha_0.0001  1177.74    0.61  0.408706  1.13858e-05 -0.000335291   
alpha_0.001   1177.74    0.61  0.408815  1.12849e-05 -0.000335602   
alpha_0.01    1177.74    0.61  0.409881  1.03374e-05 -0.000338599   
alpha_1       1193.92    0.61  0.446935  -2.2092e-06 -0.000319734   
alpha_5       1228.28    0.61  0.451007 -2.29951e-06 -0.000139688   
alpha_10       1240.3    0.61   0.44946 -1.55873e-06 -8.05609e-05   
alpha_20       1248.1    0.61  0.448124 -9.22104e-07 -4.35413e-05   

                 locxLocy shot_per_game combined_shot_type_woe  month_woe  \
alpha_1e-15  -1.09811e-08    0.00444394               0.174254   0.169172   
alpha_1e-10  -1.09811e-08    0.00444394               0.174254   0.169172   
alpha_1e-08  -1.09811e-08    0.00444394               0.174254   0.169172   
alpha_0.0001 -1.09162e-08    0.00444333               0.174235   0.169153   
alpha_0.001  -1.03365e-08    0.00443787               0.174061   0.168983   
alpha_0.01    -4.9963e-09    0.00438393               0.172346   0.167305   
alpha_1      -5.18757e-09    0.00178266              0.0872987  0.0799234   
alpha_5      -1.96207e-08   0.000474853              0.0306254  0.0254712   
alpha_10     -1.39408e-08   0.000242078              0.0169816  0.0137314   
alpha_20      -8.3933e-09   0.000121501             0.00898618  0.0071423   

             period_min_woe  
alpha_1e-15       0.0875461  
alpha_1e-10       0.0875461  
alpha_1e-08       0.0875461  
alpha_0.0001      0.0875363  
alpha_0.001       0.0874485  
alpha_0.01        0.0865843  
alpha_1           0.0457053  
alpha_5           0.0170229  
alpha_10         0.00958732  
alpha_20         0.00511987

In [44]:
from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()
lin_reg.fit(predictors, data_train_for_an['shot_made_flag'])
y_pred = lin_reg.predict(predictors)
roc_auc = np.round(roc_auc_score(data_train_for_an['shot_made_flag'], y_pred), 2)
rss = sum((y_pred-data_train_for_an['shot_made_flag']) ** 2)
print (rss, roc_auc)

4718.976236970063 0.61


In [45]:
y_pred1 = lin_reg.predict(predictors1)
roc_auc = np.round(roc_auc_score(data_test_for_an['shot_made_flag'], y_pred1), 2)
rss = sum((y_pred1-data_test_for_an['shot_made_flag']) ** 2)
print (rss, roc_auc)

1179.3343051351846 0.61


In [46]:
data_train.columns

Index(['action_type', 'combined_shot_type', 'game_event_id', 'game_id', 'lat',
       'loc_x_x', 'loc_y_x', 'lon', 'minutes_remaining', 'period', 'playoffs',
       'season', 'seconds_remaining', 'shot_distance', 'shot_made_flag',
       'shot_type', 'shot_zone_area', 'shot_zone_basic', 'shot_zone_range',
       'team_id', 'team_name', 'game_date', 'matchup', 'opponent', 'shot_id',
       'loc_x_y', 'loc_y_y', 'month', 'shot_per_game', 'shots', 'minutes_cat',
       'period_str', 'period_min', 'locxLocy'],
      dtype='object')

In [47]:

data_train_fac=data_train.drop(['action_type','period', 'game_event_id', 'game_id','season','minutes_remaining','shot_type', 'shot_zone_area', 'shot_zone_basic', 'shot_zone_range',
'team_id', 'team_name',  'matchup', 'shot_id','shot_per_game',  'game_date','minutes_cat','period_str'], axis=1)
data_test_fac=data_test.drop(['action_type','period', 'game_event_id', 'game_id','minutes_remaining','season','shot_type', 'shot_zone_area', 'shot_zone_basic', 'shot_zone_range',
       'team_id', 'team_name',  'matchup', 'shot_id',
        'shot_per_game',  'game_date','minutes_cat',
       'period_str'], axis=1)
res=['combined_shot_type', 'month', 'period_min', 'opponent']
for i in res:
    get_woe_v1(data_train_fac, data_test_fac, i, 'shot_made_flag')
predictors2=data_train_fac.drop('shot_made_flag', axis=1)
predictors2=predictors2.drop(res,axis=1)
from sklearn.decomposition import PCA
pca = PCA(n_components=predictors2.shape[1]).fit(predictors2)
for i in range(1, predictors2.shape[1] + 1):
    print("n-components=%s, explained variance %.3f" % (i, sum(pca.explained_variance_ratio_[:i])))

n-components=1, explained variance 0.999
n-components=2, explained variance 1.000
n-components=3, explained variance 1.000
n-components=4, explained variance 1.000
n-components=5, explained variance 1.000
n-components=6, explained variance 1.000
n-components=7, explained variance 1.000
n-components=8, explained variance 1.000
n-components=9, explained variance 1.000
n-components=10, explained variance 1.000
n-components=11, explained variance 1.000
n-components=12, explained variance 1.000
n-components=13, explained variance 1.000
n-components=14, explained variance 1.000
n-components=15, explained variance 1.000


In [48]:
new_factors = PCA(n_components=2).fit_transform(predictors2)
model2 = LinearRegression().fit(new_factors, data_train_fac['shot_made_flag'])
model2.score(new_factors, data_train_fac['shot_made_flag'])

0.0010395419546123108

плохо описывают

следующий этап - проанализировать решения участников на кагле. Больше всего понра решение с площадями.
https://www.kaggle.com/selfishgene/psychology-of-a-professional-athlete
Хоть там и подход несколько иной, и информации о качестве я не нашла, но истина где-то в этом
Качество других решений варьируется от 0.6 до 0.7, те манипуляции с данными не приносят существенных результатов.
Каждый делает по -своему. а качесвто одно и то же.

В том примере исследователь считает площади и процент успешных попыток.
По аналогии, но более тупо- просто посчитаем WOE по сгруппир критеряям
shot_zone_area        30697 non-null object
shot_zone_basic       30697 non-null object
shot_zone_range       30697 non-null object

In [49]:
data_train['combined']=data_train['shot_zone_area']+data_train['shot_zone_basic']+data_train['shot_zone_range']
data_test['combined']=data_test['shot_zone_area']+data_test['shot_zone_basic']+data_test['shot_zone_range']

In [50]:
columns_2=['combined_shot_type', 'month', 'period_min','shot_per_game', 'combined', 'shot_made_flag']
data_tr_an2=data_train[columns_2]
data_test_an2=data_test[columns_2]

In [51]:
columns_woe=['combined_shot_type', 'month', 'period_min', 'combined']
for i in columns_woe:
    get_woe_v1(data_tr_an2, data_test_an2, i, 'shot_made_flag')

C:\Users\volko\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # This is added back by InteractiveShellApp.init_path()
C:\Users\volko\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if sys.path[0] == '':


In [52]:
data_tr_an2=data_tr_an2.drop(columns_woe, axis=1)
data_test_an2=data_test_an2.drop(columns_woe, axis=1)

In [53]:
predictors_w=data_tr_an2.drop('shot_made_flag', axis=1)
predictors_w1=data_test_an2.drop('shot_made_flag', axis=1)

In [54]:
lin_reg_w = LinearRegression()
lin_reg_w.fit(predictors_w, data_tr_an2['shot_made_flag'])
y_pred_w = lin_reg_w.predict(predictors_w)
roc_auc = np.round(roc_auc_score(data_tr_an2['shot_made_flag'], y_pred_w), 2)
rss = sum((y_pred_w-data_tr_an2['shot_made_flag']) ** 2)
print (rss, roc_auc)

4708.916513726672 0.62


можно сказать, что не лучше. Можно было б, как в примере- выявить типы стратегий игрокам по сегментам поля и ситуациям.
Возможно, это дало бы результат.

**Задания:**

1. Провести анализ данных. Много хороших примеров анализа можно посмотреть здесь https://www.kaggle.com/c/kobe-bryant-shot-selection/kernels
2. Подготовить фичи для обучения модели - нагенерить признаков, обработать пропущенные значения, проверить на возможные выбросы, обработать категориальные признаки и др.
3. Обучить линейную модель, Lasso, Ridge на тех же признаках - построить сравнительную таблицу коэффициентов, сделать заключения о том, как меняется величина коэффициентов, какие зануляются. Посчитать RSS

**Дополнительно**
4. Сравнить результаты на тестовом наборе данных - сделать train_test_split в самом начале, подготовить переменные, сравнить результаты работы классификаторов (те же 3), метрика ROC AUC
5. Построить PCA на подготовленных признаках, посмотреть, какие компоненты составляют наибольшую часть дисперсии целевой переменной